### PR TITLE
stock_scheduler_assignation_horizon: wider settings approach

### PR DESCRIPTION
**Pull request by i-vyshnevska**
scheduler ususally running in global mode when company is not set so settings defined on company level are not applied

### CODE REVIEW

**Comment by simahawk** [comment]
s/approuche/approach
s/ususally/usually

Also, can you rephrase the description please?

File: stock_scheduler_assignation_horizon/models/res_config_settings.py [hunk 1]
```diff
@@ -27,3 +40,13 @@ def _check_moves_assignation_horizon(self):
         for record in self:
             if record.moves_assignation_horizon < 0:
                 raise ValidationError(_("The assignation horizon cannot be negative"))
+
+    def get_global_is_moves_assignation_limited(self):
+        get_param = self.env["ir.config_parameter"].sudo().get_param
+        param = "s_scheduler_assignation_horizon.global_moves_assignation"
+        return safe_eval(get_param(param, False))
```
Comment (simahawk) [inline stock_scheduler_assignation_horizon/models/res_config_settings.py]: no need for safe_eval, pls use `str2bool`

Comment (simahawk) [reply stock_scheduler_assignation_horizon/models/res_config_settings.py]: Also, these methods should not stay here. I would move them to where you need them: on the proc group.

File: stock_scheduler_assignation_horizon/models/res_config_settings.py [hunk 2]
```diff
@@ -3,11 +3,24 @@
...
 
 
 class ResConfigSettings(models.TransientModel):
     _inherit = "res.config.settings"
 
+    global_moves_assignation = fields.Boolean(
+        "Use global Assignation Limit",
+        config_parameter="s_scheduler_assignation_horizon.global_moves_assignation",
+        readonly=False,
+        help="Check this box to allow limit settings " "for each company independent",
```
Comment (simahawk) [inline stock_scheduler_assignation_horizon/models/res_config_settings.py]:
```suggestion
        help="Enable scheduler assignation horizon limit for all companies",
```

File: stock_scheduler_assignation_horizon/models/res_config_settings.py [hunk 3]
```diff
@@ -3,11 +3,24 @@
 
 from odoo import _, api, fields, models
 from odoo.exceptions import ValidationError
+from odoo.tools.safe_eval import safe_eval
 
 
 class ResConfigSettings(models.TransientModel):
     _inherit = "res.config.settings"
 
+    global_moves_assignation = fields.Boolean(
```
Comment (simahawk) [inline stock_scheduler_assignation_horizon/models/res_config_settings.py]:
```suggestion
    stock_horizon_move_assignation = fields.Boolean(
```

File: stock_scheduler_assignation_horizon/models/res_config_settings.py [hunk 4]
```diff
@@ -3,11 +3,24 @@
...
+        "Use global Assignation Limit",
+        config_parameter="s_scheduler_assignation_horizon.global_moves_assignation",
+        readonly=False,
+        help="Check this box to allow limit settings " "for each company independent",
+    )
+    global_moves_assignation_horizon = fields.Integer(
```
Comment (simahawk) [inline stock_scheduler_assignation_horizon/models/res_config_settings.py]:
```suggestion
    stock_horizon_move_assignation_limit = fields.Integer(
```

File: stock_scheduler_assignation_horizon/readme/DESCRIPTION.rst [hunk 5]
```diff
@@ -1 +1,4 @@
-This module allows to set a time horizon that will prevent the scheduler from marking stock pickings as *ready* before the indicated number of days (horizon).
+This module allows to set a time horizon that will prevent the scheduler from marking stock
+pickings as *ready* before the indicated number of days (horizon).
+Options can be enabled globally and for each company separately, remember that to enable
+mode for the company scheduler needs to be modified to set the company correctly.
```
Comment (simahawk) [inline stock_scheduler_assignation_horizon/readme/DESCRIPTION.rst]:
```suggestion

This behavior can be enabled globally via settings or for each company separately.
```
Can you please add a CONFIGURE.rst?

**Review by simahawk — CHANGES_REQUESTED** [review]

File: stock_scheduler_assignation_horizon/data/ir_config_parameter.xml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo noupdate="1">
+
+    <record id="global_is_moves_assignation_limited" model="ir.config_parameter">
+        <field
+            name="key"
+        >s_scheduler_assignation_horizon.stock_horizon_move_assignation</field>
```
Comment (ivantodorovich) [inline stock_scheduler_assignation_horizon/data/ir_config_parameter.xml]:
```suggestion
        >stock_scheduler_assignation_horizon.stock_horizon_move_assignation</field>
```
IMO better not to abbreviate this. It's for the users

File: stock_scheduler_assignation_horizon/models/procurement_group.py [hunk 7]
```diff
@@ -38,3 +44,13 @@ def _get_moves_to_assign_domain_assignation_limited(self, domain, company):
...
+
+    def get_global_is_moves_assignation_limited(self):
+        get_param = self.env["ir.config_parameter"].sudo().get_param
+        param = "s_scheduler_assignation_horizon.stock_horizon_move_assignation"
+        return str2bool(get_param(param, False))
+
+    def get_global_moves_limit_horizon(self):
+        get_param = self.env["ir.config_parameter"].sudo().get_param
+        param = "s_scheduler_assignation_horizon.stock_horizon_move_assignation_limit"
+        return safe_eval(get_param(param, 0))
```
Comment (ivantodorovich) [inline stock_scheduler_assignation_horizon/models/procurement_group.py]: If you expect an int, use `int(get_param(param, 0))`.

Maybe:

```py
        try:
            limit = int(get_param(param, 7))
        except ValueError:
            _logger.error("Wrong value for config parameter: %s", param)
            limit = 0
        return limit
```

Comment (grindtildeath) [reply stock_scheduler_assignation_horizon/models/procurement_group.py]: Nitpicking, but IMO the logger mentioned by @ivantodorovich should use a warning level and show it will use 0. e.g
```
        try:
            param_value = get_param(param, 0)
            limit = int(param_value)
        except ValueError:
            limit = 0
            _logger.warning("Wrong value for ir.config_parameter %s: %s . Using %s as default value", % (param, param_value, limit))
        return limit
```

**Comment by grindtildeath** [comment]
Instead of adding another parameter at instance level, I'm wondering if we shouldn't have added this configuration on the picking type as is the case in Odoo 15... :thinking: 

Anyway IMO we better keep only the new setting as `ir.config_parameter` and remove the old one to avoid having a setting at company that doesn't do what is expected. (and since the `procurement_jit` module is not company dependent neither, I'm fine with this choice).

For this we still need a migration script though.

File: stock_scheduler_assignation_horizon/__manifest__.py [hunk 8]
```diff
@@ -3,7 +3,7 @@
 {
     "name": "Scheduler assignation horizon",
     "summary": "Set a timeframe limit to the delivery scheduler",
-    "version": "14.0.1.0.0",
+    "version": "14.0.1.0.1",
```
Comment (simahawk) [inline stock_scheduler_assignation_horizon/__manifest__.py:6]: can you set the dev status as alpha pls?

File: stock_scheduler_assignation_horizon/migrations/14.0.1.0.1/pre-migration.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def migrate(cr, version):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    if column_exists(cr, "res_company", "is_moves_assignation_limited"):
+        query = "SELECT is_moves_assignation_limited FROM res_company;"
+        cr.execute(query)
+        res = cr.fetchall()
+
+        create_param = [True for x in res if x[0] is True]
```
Comment (simahawk) [inline stock_scheduler_assignation_horizon/migrations/14.0.1.0.1/pre-migration.py:15]: Can't you simply select one where is True?

File: stock_scheduler_assignation_horizon/models/procurement_group.py [hunk 10]
```diff
@@ -2,31 +2,33 @@
...
     _inherit = "procurement.group"
 
     def _get_moves_to_assign_domain(self, company_id):
         domain = super()._get_moves_to_assign_domain(company_id)
-        company = self.env["res.company"].browse(company_id)
-        if company.is_moves_assignation_limited:
-            domain = self._get_moves_to_assign_domain_assignation_limited(
-                domain, company
-            )
+        global_limit = self.get_global_is_moves_assignation_limited()
```
Comment (simahawk) [inline stock_scheduler_assignation_horizon/models/procurement_group.py]: Pass the company. You keep the hook for whoever wants to customize the behavior by company.

**Review by simahawk — COMMENTED** [review]
Minor remarks.

@grindtildeath @ivan can you update your reviews?

File: stock_scheduler_assignation_horizon/models/procurement_group.py [hunk 11]
```diff
@@ -38,3 +40,24 @@ def _get_moves_to_assign_domain_assignation_limited(self, domain, company):
                 [("date", "<=", max_date)],
             ]
         )
+
+    def get_global_is_moves_assignation_limited(self):
```
Comment (ivantodorovich) [inline stock_scheduler_assignation_horizon/models/procurement_group.py]:
```suggestion
    def _get_global_is_moves_assignation_limited(self):
```
Should be private

File: stock_scheduler_assignation_horizon/models/procurement_group.py [hunk 12]
```diff
@@ -38,3 +40,24 @@ def _get_moves_to_assign_domain_assignation_limited(self, domain, company):
...
+        get_param = self.env["ir.config_parameter"].sudo().get_param
+        param = "stock_scheduler_assignation_horizon.stock_horizon_move_assignation"
+        return str2bool(get_param(param, False))
+
+    def get_global_moves_limit_horizon(self):
```
Comment (ivantodorovich) [inline stock_scheduler_assignation_horizon/models/procurement_group.py]:
```suggestion
    def _get_global_moves_limit_horizon(self):
```
Should be private

File: stock_scheduler_assignation_horizon/migrations/14.0.1.0.1/pre-migration.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
+# Copyright 2022 Camtocamp
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import SUPERUSER_ID, api
+from odoo.tools.sql import column_exists
+
+
+def migrate(cr, version):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    if column_exists(cr, "res_company", "is_moves_assignation_limited"):
```
Comment (ivantodorovich) [inline stock_scheduler_assignation_horizon/migrations/14.0.1.0.1/pre-migration.py:10]: You can count on the column to exist, because it did before this new module version.
If the migration is executed, it means you're coming from a previous version: the column definitely exists

In any case, it doesn't hurt to check I guess.

I would avoid the nested `if` indentation by reverting your conditions.

```suggestion
    if not column_exists(cr, "res_company", "is_moves_assignation_limited"):
        return
    # ...
```

Alternatively, since the field already exist.. I think you can simply use the env rather than sql queries

```py
    if "is_moves_assignation_limited" not in env["res.company"]._fields:
        return
    companies = env["res.company"].search([])

    stock_horizon_move_assignation = any(c.is_moves_assignation_limited for c in companies)
    if not stock_horizon_move_assignation:
        return
    _create_ir_config_param(
        "stock_scheduler_assignation_horizon.stock_horizon_move_assignation",
        True,
        "stock_scheduler_assignation_horizon.is_moves_assignation_limited",
    )

    stock_horizon_move_assignation_limit = min(
        c.moves_assignation_horizon for c in companies
        if c.moves_assignation_horizon,
        default=0
    )
    if not stock_horizon_move_assignation_limit:
        return
    _create_ir_config_param(
       "stock_scheduler_assignation_horizon.stock_horizon_move_assignation_limit",
        stock_horizon_move_assignation_limit,
        "stock_scheduler_assignation_horizon.stock_scheduler_assignation_horizon",
    )

def _create_ir_config_param(key, value, xmlid):
    # ...
```

File: stock_scheduler_assignation_horizon/models/procurement_group.py [hunk 14]
```diff
@@ -2,31 +2,33 @@
...
+        global_limit = self.get_global_is_moves_assignation_limited()
+        if global_limit:
+            delay = self.get_global_moves_limit_horizon()
+            domain = self._get_moves_to_assign_domain_assignation_limited(domain, delay)
         return domain
 
-    def _get_moves_to_assign_domain_assignation_limited(self, domain, company):
-        company_tz = pytz.timezone(company.partner_id.tz or "UTC")
+    def _get_moves_to_assign_domain_assignation_limited(self, domain, delay_days):
+        applied_tz = pytz.timezone("UTC")
```
Comment (grindtildeath) [inline stock_scheduler_assignation_horizon/models/procurement_group.py]: IMO no need to use a timezone to convert from UTC to UTC :upside_down_face: 

I would say we can define the max_date as `datetime.now() + timedelta(days=delay_days)` as you did, without  having to combine with max time.

File: stock_scheduler_assignation_horizon/models/procurement_group.py [hunk 15]
```diff
@@ -2,31 +2,33 @@
...
         domain = super()._get_moves_to_assign_domain(company_id)
-        company = self.env["res.company"].browse(company_id)
-        if company.is_moves_assignation_limited:
-            domain = self._get_moves_to_assign_domain_assignation_limited(
-                domain, company
-            )
+        global_limit = self.get_global_is_moves_assignation_limited()
+        if global_limit:
+            delay = self.get_global_moves_limit_horizon()
+            domain = self._get_moves_to_assign_domain_assignation_limited(domain, delay)
```
Comment (grindtildeath) [inline stock_scheduler_assignation_horizon/models/procurement_group.py]: While we're in a big refactor, could we combine the domain here instead of inside the function? It would allow for easier override doing this:
```
horizon_domain = self._get_assignation_horizon_domain(delay)
domain = AND([domain, horizon_domain])
```

File: stock_scheduler_assignation_horizon/models/procurement_group.py [hunk 16]
```diff
@@ -2,31 +2,33 @@
...
         return domain
 
-    def _get_moves_to_assign_domain_assignation_limited(self, domain, company):
-        company_tz = pytz.timezone(company.partner_id.tz or "UTC")
+    def _get_moves_to_assign_domain_assignation_limited(self, domain, delay_days):
```
Comment (grindtildeath) [inline stock_scheduler_assignation_horizon/models/procurement_group.py]:
```suggestion
    def _get_assignation_horizon_domain(self, delay_days):
```
IMO we can use a shorter function name :sweat_smile:

File: stock_scheduler_assignation_horizon/models/procurement_group.py (hunk 11 above)
Comment (grindtildeath) [reply stock_scheduler_assignation_horizon/models/procurement_group.py]: Since we don't have any company settings anymore, a shorter version as `_is_assignation_horizon_limited` would also be nice :innocent:

File: stock_scheduler_assignation_horizon/models/procurement_group.py (hunk 12 above)
Comment (grindtildeath) [reply stock_scheduler_assignation_horizon/models/procurement_group.py]: `_get_assignation_horizon_limit` ? :innocent:

**Review by grindtildeath — COMMENTED** [review]
Small things, mostly naming of things that can be improved to make the code more readable.

File: stock_scheduler_assignation_horizon/models/procurement_group.py [hunk 17]
```diff
@@ -2,39 +2,48 @@
...
-        return expression.AND(
-            [
-                domain,
-                [("date", "<=", max_date)],
-            ]
+    def _get_assignation_horizon_domain(self, delay_days):
+        max_date = datetime.now() + timedelta(days=delay_days)
+        return [("date", "<=", max_date)]
+
+    def _get_global_is_moves_assignation_limited(self):
```
Comment (grindtildeath) [inline stock_scheduler_assignation_horizon/models/procurement_group.py]:
```suggestion
    def _is_assignation_horizon_limited(self):
```
As asked in previous comment :wink:

File: stock_scheduler_assignation_horizon/models/procurement_group.py [hunk 18]
```diff
@@ -2,39 +2,48 @@
...
     _inherit = "procurement.group"
 
     def _get_moves_to_assign_domain(self, company_id):
         domain = super()._get_moves_to_assign_domain(company_id)
-        company = self.env["res.company"].browse(company_id)
-        if company.is_moves_assignation_limited:
-            domain = self._get_moves_to_assign_domain_assignation_limited(
-                domain, company
-            )
+        global_limit = self._get_global_is_moves_assignation_limited()
```
Comment (grindtildeath) [inline stock_scheduler_assignation_horizon/models/procurement_group.py]:
```suggestion
        global_limit = self._is_assignation_horizon_limited()
```

**Review by grindtildeath — APPROVED** [review]
Can you please check the comments on the migration script? Pre approving anyway.

**Comment by grindtildeath** [comment]
@simahawk Can you please update your review and remove stale label?

**Review by rousseldenis — APPROVED** [review]
Code review

**Review by simahawk — APPROVED** [review]
